### ISSUE TITLE
How do we make changes to the DOM elements generated by the inspector? Currently having issue identifying selectors in the DOM that are resilient to change. Wondering how other teams are managing this?

### ISSUE THREAD

**Issue by richardforjoe** [rdkcentral/Lightning#260]
Issue: so using the TicTacToe game as an example. When asserting that the FocusIndicator is aligned to the correct menu item. I noticed in the DOM the div element for the FocusIndicator was completely separate to the Item it was associated to via the UI Menu. See screenshot attached

<img width="1305" alt="Screenshot 2020-11-18 at 16 44 50" src="https://user-images.githubusercontent.com/54558593/99723199-c18cc580-2aa9-11eb-8d77-1c14e524e2b9.png">

i've had to use the y attribute to check if a menu item is in focus or not. As y is a co-ordinate, soon as that changes the test will break. How do we add attributes to the DOM and how can associated components in the Dom be linked either via the DOM tree or attributes eg FocusIndicator being a child of the Item in focus or an attribute on the Item indicating its in focus. 

PS: This is just one example. Main issue is how do we manipulate the DOM to provide more reliable locators for automation.

`Parent <div ref=“Items”
child <div type=“Item” texture-text=“Menu…
child <div type=“Item” texture-text=“Menu…
child <div type=“Item” texture-text=“Menu…
child <div type=“Item” texture-text=“Menu…
child <div ref=“FocusIndicator” y=“ “
Workaround to check if menu has an focus indicator or not
static assertFocusIndicator(MenuItem)
{
 let focus;
 switch(MenuItem) {
  case "Start New Game":
   focus = "5"
   break;
  case "Continue":
   focus = "95"
   break;
  case "About":
   focus = "185"
   break;
  case "Exit":
   focus = "275"
   break;
 }
  return cy.get(FOCUS_INDICATOR_COMPONENT).invoke('attr', 'y').should('eq', `${focus}`)
}`

**Comment by g-zachar** [comment]
Hi @richardforjoe 

could you point me to example you're using?

Best regards

**Comment by richardforjoe** [comment]
Hi @g-zachar , Sure

The repo i'm using is here: https://github.com/richardforjoe/lightning

Thank you
Richard

**Comment by g-zachar** [comment]
@richardforjoe 

The `FocusIndicator` you're using is element specific to this particular app (check `Menu.js` source code). The HTML DOM is generated by Lightning inspector ([link](https://github.com/rdkcentral/Lightning/blob/master/devtools/lightning-inspect.js)), by default lightning app operates without it. It is one way binding, mostly used for debugging purposes, altering it will not change the app itself in any way.

There is no definitive way to do tests for Lightning apps, though there are several ideas floating around (e.g. [jest snapshots](https://github.com/rdkcentral/Lightning/commit/1790cb356465455c3e8b04f87cdd62392660f353). You can read app's layout in pretty readable format using aforementioned methods, bootstrapping test app should be no different from how Lightning is bootstrapped for production, there is additional step of integrating it into your test framework. Any manipulation could be achieved by either using Lightning's API directly or through event handlers (e.g. for navigating the app).

As for locators, you can request changes to the source of inspector here, or you could even modify/rewrite it for your own purposes as it is just simple loosely coupled overlay on top of Lightning.

Best regards

**Comment by richardforjoe** [comment]
Awesome @g-zachar Thanks a lot for this. I'l spend some time digesting this, very helpful. 

Another quick question, do you have any details on accessing the lightning api directly? I'm assuming it's like a rest endpoint or do you mean some sort of interface thats only available to the code?

Many Thanks
Richard

**Comment by g-zachar** [comment]
@richardforjoe 

No, by API I meant same JavaScript API that is used for writing apps can be used for writing tests.

Best Regards

**Comment by andriichumak** [comment]
There was a pull request with changes, that should make static selectors possible: https://github.com/rdkcentral/Lightning/pull/288 ;)
